File: bc2025/src/philip_06/roles/Splasher.cpp
#include "roles/Splasher.hpp"

#include <algorithm>
#include <cmath>
#include <iostream>

#include "Helper/Comms.hpp"
#include "RobotPlayer.hpp"
#include "roles/Mopper.hpp"
#include "utils/Utils.hpp"

namespace philip_06 {
namespace roles {

using philip_06::Helper::Comms;
using philip_06::utils::Utils;

Random* Splasher::rng = nullptr;
bool Splasher::type = false;
MapLocation Splasher::destination;  // Java: null
int Splasher::opponentPaintSquares = 0;
std::unordered_set<MapLocation> Splasher::paintTowers;
int Splasher::turnsSinceLastAttack = -100;
bool Splasher::targetRuinsOnly = false;
MapLocation Splasher::oppSpawn;  // Java: null

void Splasher::retreat(RobotController rc) {
    MapLocation bestOpt;  // Java: null
    for (const MapLocation& curr : paintTowers) {
        if (bestOpt.isNull() ||
            (rc.getLocation().distanceSquaredTo(curr) < rc.getLocation().distanceSquaredTo(bestOpt) &&
             (!rc.canSenseLocation(bestOpt) || !rc.senseRobotAtLocation(bestOpt).getLocation().isNull()))) {
            bestOpt = curr;
        }
    }
    if (bestOpt.isNull()) {
        bestOpt = getNextCorner(rc);
        Mopper::moveTowardsMindfully(rc, bestOpt);
        return;
    }
    int transfer = paintCapacity(rc.getType()) - rc.getPaint();
    if (rc.canSenseLocation(bestOpt) && !rc.senseRobotAtLocation(bestOpt).getLocation().isNull()) {
        transfer = std::min(transfer, rc.senseRobotAtLocation(bestOpt).getPaintAmount());
        rc.setIndicatorString(std::to_string(transfer));
        if (rc.canTransferPaint(bestOpt, -transfer)) rc.transferPaint(bestOpt, -transfer);
    }
    Mopper::moveTowardsMindfully(rc, bestOpt);
}

void Splasher::bestRuinAttack(RobotController rc) {
    MapLocation closestRuin = Utils::closestRuin(rc);
    (void)closestRuin;  // Java: assigned but unused in this method
    bool forced = false;
    int best = -1000;
    MapLocation target;  // Java: null
    vector<MapInfo> nearby = rc.senseNearbyMapInfos(4);
    int xavg = 0, yavg = 0;

    for (const MapInfo& loc : rc.senseNearbyMapInfos(-1)) {
        if (isEnemy(loc.getPaint())) {
            opponentPaintSquares++;
            xavg += loc.getMapLocation().x;
            yavg += loc.getMapLocation().y;
        }
    }
    if (opponentPaintSquares != 0) {
        // Java Math.round(float) == floor(x + 0.5f)
        xavg = (int)std::floor((float)xavg / opponentPaintSquares + 0.5f);
        yavg = (int)std::floor((float)yavg / opponentPaintSquares + 0.5f);
        forced = rc.getLocation().isWithinDistanceSquared(MapLocation(xavg, yavg), 1);
    }
    /*
    int[] temp = new int[nearby.length];
    for(int i = 0; i<temp.length; i++){
        temp[i] = i;
    }
    Utils.shuffleArray(temp);
    */
    for (const MapInfo& loc : nearby) {
        int damage = 0;
        bool prim = false, sec = false;
        for (const MapInfo& hit : rc.senseNearbyMapInfos(loc.getMapLocation(), 4)) {
            // 1 for paint, 3 for enemy
            if (hit.isWall()) continue;
            if (isAlly(hit.getPaint())) {
                damage -= 2;
                if (hit.getPaint() == PaintType::ALLY_PRIMARY) {
                    prim = true;
                    if (sec)
                        break;
                }
                if (hit.getPaint() == PaintType::ALLY_SECONDARY) {
                    sec = true;
                    if (prim)
                        break;
                }
            }
            else if (hit.hasRuin() && (rc.senseRobotAtLocation(hit.getMapLocation()).getLocation().isNull() ||
                                       rc.senseRobotAtLocation(hit.getMapLocation()).getTeam() != rc.getTeam()))
                damage += 10000;
            else damage += 1;

        }
        if (damage >= best) {
            best = damage;
            target = loc.getMapLocation();
            if (sec)
                type = true;
            if (prim)
                type = false;
        }
    }
    if (opponentPaintSquares == 0)
        turnsSinceLastAttack = 0;
    if (best > 10000 || forced) {
        if (rc.canAttack(target)) {
            turnsSinceLastAttack = 0;
            rc.attack(target, type);
            Mopper::explore(rc);
        } else {
            turnsSinceLastAttack++;
            Mopper::explore(rc);
        }
    } else if (opponentPaintSquares != 0) {
        turnsSinceLastAttack++;
        Mopper::explore(rc);
    } else {
        turnsSinceLastAttack++;
        Mopper::explore(rc);
    }
}

void Splasher::bestAttack(RobotController rc) {
    MapLocation closestRuin = Utils::closestRuin(rc);
    bool forced = false;
    int best = -1000;
    MapLocation target;  // Java: null
    vector<MapInfo> nearby = rc.senseNearbyMapInfos(4);
    int xavg = 0, yavg = 0;

    for (const MapInfo& loc : rc.senseNearbyMapInfos(-1)) {
        if (loc.isResourcePatternCenter() && isEnemy(loc.getPaint())) {
            opponentPaintSquares += 10;
            xavg += 10 * loc.getMapLocation().x;
            yavg += 10 * loc.getMapLocation().y;
        } else if (isEnemy(loc.getPaint())) {
            opponentPaintSquares++;
            xavg += loc.getMapLocation().x;
            yavg += loc.getMapLocation().y;
        }
    }
    if (opponentPaintSquares != 0) {
        // Java Math.round(float) == floor(x + 0.5f)
        xavg = (int)std::floor((float)xavg / opponentPaintSquares + 0.5f);
        yavg = (int)std::floor((float)yavg / opponentPaintSquares + 0.5f);
        forced = rc.getLocation().isWithinDistanceSquared(MapLocation(xavg, yavg), 1);
    }
    /*
    int[] temp = new int[nearby.length];
    for(int i = 0; i<temp.length; i++){
        temp[i] = i;
    }
    Utils.shuffleArray(temp);
    */
    for (const MapInfo& loc : nearby) {
        int damage = 0;
        bool prim = false, sec = false;
        for (const MapInfo& hit : rc.senseNearbyMapInfos(loc.getMapLocation(), 4)) {
            // 1 for paint, 3 for enemy
            if (hit.isWall() || hit.hasRuin()) continue;
            if (isAlly(hit.getPaint())) {
                damage--;
                if (hit.getPaint() == PaintType::ALLY_PRIMARY) {
                    prim = true;
                    if (sec)
                        break;
                }
                if (hit.getPaint() == PaintType::ALLY_SECONDARY) {
                    sec = true;
                    if (prim)
                        break;
                }
            }
            else if (hit.getPaint() == PaintType::EMPTY) damage += 1;
            else if (hit.getMapLocation().distanceSquaredTo(loc.getMapLocation()) <= 2) {
                if (hit.isResourcePatternCenter() && isEnemy(hit.getPaint()))
                    damage += 30;
                else if (!closestRuin.isNull() && closestRuin.isWithinDistanceSquared(hit.getMapLocation(), 8))
                    damage += 5;
                else
                    damage += 3;
            }
        }
        if (damage >= best) {
            best = damage;
            target = loc.getMapLocation();
            if (sec)
                type = true;
            if (prim)
                type = false;
        }
    }
    if (opponentPaintSquares == 0)
        turnsSinceLastAttack = 0;
    if (best > 11 || forced) {
        if (rc.canAttack(target)) {
            turnsSinceLastAttack = 0;
            rc.attack(target, type);
            if (opponentPaintSquares != 0 &&
                rc.canMove(rc.getLocation().directionTo(MapLocation(xavg, yavg)).opposite()))
                rc.move(rc.getLocation().directionTo(MapLocation(xavg, yavg)).opposite());
        } else {
            turnsSinceLastAttack++;
            //Mopper.explore(rc);
        }
    } else if (opponentPaintSquares != 0) {
        turnsSinceLastAttack++;
        Mopper::moveTowardsMindfully(rc, MapLocation(xavg, yavg));
    } else {
        turnsSinceLastAttack++;
        //Mopper.explore(rc);
    }
}

Direction Splasher::shouldMove(RobotController rc, Direction dir) {
    if (!rc.canMove(dir)) return Direction::all[Direction::CENTER];  // Java: null
    vector<MapLocation> nearby = rc.senseNearbyRuins(16);
    for (const MapLocation& loc : nearby) {
        if (!canSenseRobotAtLocation(rc, loc)) continue;
        RobotInfo tower = rc.senseRobotAtLocation(loc);
        if (tower.getTeam() == rc.getTeam()) continue;
        dir = rc.getLocation().directionTo(loc).opposite();
    }
    if (rc.canSenseLocation(rc.getLocation().add(dir)) &&
        !isEnemy(rc.senseMapInfo(rc.getLocation().add(dir)).getPaint()))
        return dir;
    dir = dir.rotateRight();
    if (rc.canSenseLocation(rc.getLocation().add(dir)) &&
        !isEnemy(rc.senseMapInfo(rc.getLocation().add(dir)).getPaint()))
        return dir;
    dir = dir.rotateLeft();
    dir = dir.rotateLeft();
    if (rc.canSenseLocation(rc.getLocation().add(dir)) &&
        !isEnemy(rc.senseMapInfo(rc.getLocation().add(dir)).getPaint()))
        return dir;
    return Direction::all[Direction::CENTER];  // Java: null
}

MapLocation Splasher::getNextCorner(RobotController rc) {
    if (rng == nullptr) rng = &RobotPlayer::rng;
    int x = rng->nextInt(3) * (rc.getMapWidth() / 2), y = rng->nextInt(3) * (rc.getMapHeight() / 2);
    x = std::max(0, std::min(rc.getMapWidth() - 1, x + rng->nextInt() % 3));
    y = std::max(0, std::min(rc.getMapHeight() - 1, y + rng->nextInt() % 3));
    return MapLocation(x, y);
}

void Splasher::run(RobotController rc) {
    if (rc.getRoundNum() < 4 || rc.getID() % 20 == 0)
        targetRuinsOnly = true;
    if (targetRuinsOnly) {
        bestRuinAttack(rc);
        return;
    }
    if (rng == nullptr) rng = &RobotPlayer::rng;
    for (const MapLocation& loc : rc.senseNearbyRuins(-1)) {
        RobotInfo tower = rc.senseRobotAtLocation(loc);
        if (tower.getLocation().isNull() || tower.getTeam() == opponent(rc.getTeam())) {
            if (!tower.getLocation().isNull())
                RobotPlayer::lastEnemyTower = tower.getLocation();
            paintTowers.erase(loc);
            continue;
        }
        if (tower.getPaintAmount() >= 150 || paintPerTurn(tower.getType()) > 0)
            paintTowers.insert(loc);
        else
            paintTowers.erase(loc);
    }

    Comms::sendMsg(rc);
    Comms::getMsg(rc);
    if (rc.getPaint() < 10) {
        for (const RobotInfo& robot : rc.senseNearbyRobots(2)) {
            if (robot.getTeam() == rc.getTeam() && isRobotType(robot.getType()) &&
                robot.getType() != UnitType::MOPPER && robot.getPaintAmount() > 50) {
                std::cout << "goodbye" << '\n';
                rc.disintegrate();
            }
        }
    }
    if (rc.getPaint() < 50) {
        retreat(rc);
        rc.setIndicatorString("retreat");
        return;
    }
    opponentPaintSquares = 0;
    //if (oppSpawn == null) destination = oppSpawn = new MapLocation(rc.getMapWidth()-rc.getLocation().x,rc.getMapHeight()-rc.getLocation().y);
    /*
    int pri = -1;
    for (Message mess : rc.readMessages(-1)) {
        int[] action = Comms.decode(mess.getBytes());

        // Only care about recent messages
        if (rc.getRoundNum() - action[4] > 50) continue;

        // action[5] = 0 => attack location
        if (action[5] == 0) {
            if (action[1] > pri) {
                pri = action[1];
                destination = new MapLocation(action[2], action[3]);
            }
        }
    }*/

    if (destination.isNull() || rc.getLocation().distanceSquaredTo(destination) < 9 ||
        rc.getRoundNum() % 69 == 0)
        destination = getNextCorner(rc);
    bestAttack(rc);
    if (opponentPaintSquares == 0) {
        Mopper::moveTowardsMindfully(rc, destination);
    }
    if (rc.isMovementReady()) {
        MapLocation best;  // Java: null
        for (const MapInfo& loc : rc.senseNearbyMapInfos(-1)) {
            if (isAlly(loc.getPaint()) &&
                (best.isNull() || rc.getLocation().distanceSquaredTo(loc.getMapLocation()) <
                                      rc.getLocation().distanceSquaredTo(best))) {
                best = loc.getMapLocation();
            }
        }
        if (!best.isNull()) {
            Mopper::moveTowardsMindfully(rc, best);
        }
    }
}

}  // namespace roles
}  // namespace philip_06

File: py/bc2025/src/philip_06/roles/Splasher.py
# Port of philip_06/roles/Splasher.java. See py/CONVENTIONS.md.
import math

from api import *
import RobotPlayer
from roles import Mopper
from utils import Utils
from Helper import Comms

# Java: import static philip_06.Helper.Comms.getMsg
getMsg = Comms.Comms.getMsg
# Java: import static philip_06.Helper.Comms.sendMsg
sendMsg = Comms.Comms.sendMsg


class Splasher:
    rng = None
    type = False
    destination = None
    opponentPaintSquares = 0
    paintTowers = set()
    turnsSinceLastAttack = -100

    @staticmethod
    def retreat(rc):
        bestOpt = None
        for curr in Splasher.paintTowers:
            if bestOpt is None or (rc.getLocation().distanceSquaredTo(curr) < rc.getLocation().distanceSquaredTo(bestOpt) and (not rc.canSenseLocation(bestOpt) or rc.senseRobotAtLocation(bestOpt) is not None)):
                bestOpt = curr
        if bestOpt is None:
            bestOpt = Splasher.getNextCorner(rc)
            Mopper.Mopper.moveTowardsMindfully(rc, bestOpt)
            return
        transfer = rc.getType().paintCapacity - rc.getPaint()
        if rc.canSenseLocation(bestOpt) and rc.senseRobotAtLocation(bestOpt) is not None:
            transfer = min(transfer, rc.senseRobotAtLocation(bestOpt).getPaintAmount())
            rc.setIndicatorString(str(transfer))
            if rc.canTransferPaint(bestOpt, -transfer): rc.transferPaint(bestOpt, -transfer)
        Mopper.Mopper.moveTowardsMindfully(rc, bestOpt)

    targetRuinsOnly = False

    @staticmethod
    def bestRuinAttack(rc):
        closestRuin = Utils.Utils.closestRuin(rc)  # Java: assigned but unused in this method
        forced = False
        best = -1000
        target = None
        nearby = rc.senseNearbyMapInfos(4)
        xavg = 0
        yavg = 0

        for loc in rc.senseNearbyMapInfos(-1):
            if loc.getPaint().isEnemy():
                Splasher.opponentPaintSquares += 1
                xavg += loc.getMapLocation().x
                yavg += loc.getMapLocation().y
        if Splasher.opponentPaintSquares != 0:
            # Java Math.round(float) == floor(x + 0.5f)
            xavg = int(math.floor(xavg / Splasher.opponentPaintSquares + 0.5))
            yavg = int(math.floor(yavg / Splasher.opponentPaintSquares + 0.5))
            forced = rc.getLocation().isWithinDistanceSquared(MapLocation(xavg, yavg), 1)
        #
        # int[] temp = new int[nearby.length];
        # for(int i = 0; i<temp.length; i++){
        #     temp[i] = i;
        # }
        # Utils.shuffleArray(temp);
        #
        for loc in nearby:
            damage = 0
            prim = False
            sec = False
            for hit in rc.senseNearbyMapInfos(loc.getMapLocation(), 4):
                # 1 for paint, 3 for enemy
                if hit.isWall(): continue
                if hit.getPaint().isAlly():
                    damage -= 2
                    if hit.getPaint() == PaintType.ALLY_PRIMARY:
                        prim = True
                        if sec:
                            break
                    if hit.getPaint() == PaintType.ALLY_SECONDARY:
                        sec = True
                        if prim:
                            break
                elif hit.hasRuin() and (rc.senseRobotAtLocation(hit.getMapLocation()) is None or rc.senseRobotAtLocation(hit.getMapLocation()).getTeam() != rc.getTeam()):
                    damage += 10000
                else:
                    damage += 1

            if damage >= best:
                best = damage
                target = loc.getMapLocation()
                if sec:
                    Splasher.type = True
                if prim:
                    Splasher.type = False
        if Splasher.opponentPaintSquares == 0:
            Splasher.turnsSinceLastAttack = 0
        if best > 10000 or forced:
            if rc.canAttack(target):
                Splasher.turnsSinceLastAttack = 0
                rc.attack(target, Splasher.type)
                Mopper.Mopper.explore(rc)
            else:
                Splasher.turnsSinceLastAttack += 1
                Mopper.Mopper.explore(rc)
        elif Splasher.opponentPaintSquares != 0:
            Splasher.turnsSinceLastAttack += 1
            Mopper.Mopper.explore(rc)
        else:
            Splasher.turnsSinceLastAttack += 1
            Mopper.Mopper.explore(rc)

    @staticmethod
    def bestAttack(rc):
        closestRuin = Utils.Utils.closestRuin(rc)
        forced = False
        best = -1000
        target = None
        nearby = rc.senseNearbyMapInfos(4)
        xavg = 0
        yavg = 0

        for loc in rc.senseNearbyMapInfos(-1):
            if loc.isResourcePatternCenter() and loc.getPaint().isEnemy():
                Splasher.opponentPaintSquares += 10
                xavg += 10 * loc.getMapLocation().x
                yavg += 10 * loc.getMapLocation().y
            elif loc.getPaint().isEnemy():
                Splasher.opponentPaintSquares += 1
                xavg += loc.getMapLocation().x
                yavg += loc.getMapLocation().y
        if Splasher.opponentPaintSquares != 0:
            # Java Math.round(float) == floor(x + 0.5f)
            xavg = int(math.floor(xavg / Splasher.opponentPaintSquares + 0.5))
            yavg = int(math.floor(yavg / Splasher.opponentPaintSquares + 0.5))
            forced = rc.getLocation().isWithinDistanceSquared(MapLocation(xavg, yavg), 1)
        #
        # int[] temp = new int[nearby.length];
        # for(int i = 0; i<temp.length; i++){
        #     temp[i] = i;
        # }
        # Utils.shuffleArray(temp);
        #
        for loc in nearby:
            damage = 0
            prim = False
            sec = False
            for hit in rc.senseNearbyMapInfos(loc.getMapLocation(), 4):
                # 1 for paint, 3 for enemy
                if hit.isWall() or hit.hasRuin(): continue
                if hit.getPaint().isAlly():
                    damage -= 1
                    if hit.getPaint() == PaintType.ALLY_PRIMARY:
                        prim = True
                        if sec:
                            break
                    if hit.getPaint() == PaintType.ALLY_SECONDARY:
                        sec = True
                        if prim:
                            break
                elif hit.getPaint() == PaintType.EMPTY:
                    damage += 1
                elif hit.getMapLocation().distanceSquaredTo(loc.getMapLocation()) <= 2:
                    if hit.isResourcePatternCenter() and hit.getPaint().isEnemy():
                        damage += 30
                    elif closestRuin is not None and closestRuin.isWithinDistanceSquared(hit.getMapLocation(), 8):
                        damage += 5
                    else:
                        damage += 3
            if damage >= best:
                best = damage
                target = loc.getMapLocation()
                if sec:
                    Splasher.type = True
                if prim:
                    Splasher.type = False
        if Splasher.opponentPaintSquares == 0:
            Splasher.turnsSinceLastAttack = 0
        if best > 11 or forced:
            if rc.canAttack(target):
                Splasher.turnsSinceLastAttack = 0
                rc.attack(target, Splasher.type)
                if Splasher.opponentPaintSquares != 0 and rc.canMove(rc.getLocation().directionTo(MapLocation(xavg, yavg)).opposite()):
                    rc.move(rc.getLocation().directionTo(MapLocation(xavg, yavg)).opposite())
            else:
                Splasher.turnsSinceLastAttack += 1
                #Mopper.explore(rc);
        elif Splasher.opponentPaintSquares != 0:
            Splasher.turnsSinceLastAttack += 1
            Mopper.Mopper.moveTowardsMindfully(rc, MapLocation(xavg, yavg))
        else:
            Splasher.turnsSinceLastAttack += 1
            #Mopper.explore(rc);

    @staticmethod
    def shouldMove(rc, dir):
        if not rc.canMove(dir): return None
        nearby = rc.senseNearbyRuins(16)
        for loc in nearby:
            if not rc.canSenseRobotAtLocation(loc): continue
            tower = rc.senseRobotAtLocation(loc)
            if tower.getTeam() == rc.getTeam(): continue
            dir = rc.getLocation().directionTo(loc).opposite()
        if rc.canSenseLocation(rc.getLocation().add(dir)) and not rc.senseMapInfo(rc.getLocation().add(dir)).getPaint().isEnemy():
            return dir
        dir = dir.rotateRight()
        if rc.canSenseLocation(rc.getLocation().add(dir)) and not rc.senseMapInfo(rc.getLocation().add(dir)).getPaint().isEnemy():
            return dir
        dir = dir.rotateLeft()
        dir = dir.rotateLeft()
        if rc.canSenseLocation(rc.getLocation().add(dir)) and not rc.senseMapInfo(rc.getLocation().add(dir)).getPaint().isEnemy():
            return dir
        return None

    oppSpawn = None

    @staticmethod
    def getNextCorner(rc):
        if Splasher.rng is None: Splasher.rng = RobotPlayer.RobotPlayer.rng
        x = Splasher.rng.nextInt(3) * (rc.getMapWidth() // 2)
        y = Splasher.rng.nextInt(3) * (rc.getMapHeight() // 2)
        x = max(0, min(rc.getMapWidth() - 1, x + Splasher.rng.nextInt() % 3))
        y = max(0, min(rc.getMapHeight() - 1, y + Splasher.rng.nextInt() % 3))
        return MapLocation(x, y)

    @staticmethod
    def run(rc):
        if rc.getRoundNum() < 4 or rc.getID() % 20 == 0:
            Splasher.targetRuinsOnly = True
        if Splasher.targetRuinsOnly:
            Splasher.bestRuinAttack(rc)
            return
        if Splasher.rng is None: Splasher.rng = RobotPlayer.RobotPlayer.rng
        for loc in rc.senseNearbyRuins(-1):
            tower = rc.senseRobotAtLocation(loc)
            if tower is None or tower.getTeam() == opponent(rc.getTeam()):
                if tower is not None:
                    RobotPlayer.RobotPlayer.lastEnemyTower = tower.getLocation()
                Splasher.paintTowers.discard(loc)
                continue
            if tower.getPaintAmount() >= 150 or tower.getType().paintPerTurn > 0:
                Splasher.paintTowers.add(loc)
            else:
                Splasher.paintTowers.discard(loc)

        sendMsg(rc)
        getMsg(rc)
        if rc.getPaint() < 10:
            for robot in rc.senseNearbyRobots(2):
                if robot.getTeam() == rc.getTeam() and robot.getType().isRobotType() and robot.getType() != UnitType.MOPPER and robot.getPaintAmount() > 50:
                    print("goodbye")
                    rc.disintegrate()
        if rc.getPaint() < 50:
            Splasher.retreat(rc)
            rc.setIndicatorString("retreat")
            return
        Splasher.opponentPaintSquares = 0
        #if (oppSpawn == null) destination = oppSpawn = new MapLocation(rc.getMapWidth()-rc.getLocation().x,rc.getMapHeight()-rc.getLocation().y);
        #
        # int pri = -1;
        # for (Message mess : rc.readMessages(-1)) {
        #     int[] action = Comms.decode(mess.getBytes());
        #
        #     // Only care about recent messages
        #     if (rc.getRoundNum() - action[4] > 50) continue;
        #
        #     // action[5] = 0 => attack location
        #     if (action[5] == 0) {
        #         if (action[1] > pri) {
        #             pri = action[1];
        #             destination = new MapLocation(action[2], action[3]);
        #         }
        #     }
        # }

        if Splasher.destination is None or rc.getLocation().distanceSquaredTo(Splasher.destination) < 9 or rc.getRoundNum() % 69 == 0:
            Splasher.destination = Splasher.getNextCorner(rc)
        Splasher.bestAttack(rc)
        if Splasher.opponentPaintSquares == 0:
            Mopper.Mopper.moveTowardsMindfully(rc, Splasher.destination)
        if rc.isMovementReady():
            best = None
            for loc in rc.senseNearbyMapInfos(-1):
                if loc.getPaint().isAlly() and (best is None or rc.getLocation().distanceSquaredTo(loc.getMapLocation()) < rc.getLocation().distanceSquaredTo(best)):
                    best = loc.getMapLocation()
            if best is not None:
                Mopper.Mopper.moveTowardsMindfully(rc, best)
